src/visitor.ts

```diff
@@ -164,12 +186,64 @@ export class CelVisitor
     return result[index]
   }
 
+  mapExpression(ctx: MapExpressionCstChildren) {
+    const mapExpression: { [key: string]: unknown } = {}
+    if (!ctx.keyValues) {
+      return {}
+    }
+    let valueType: string = ''
+    for (const keyValuePair of ctx.keyValues) {
+      const [key, value] = this.visit(keyValuePair)
+      if (valueType === '') {
+        valueType = getCelType(value)
+      }
+      if (getCelType(key) != CelType.string) {
+        throw new CelEvaluationError(`invalid_argument: ${key}`)
+      }
+      if (valueType !== getCelType(value)) {
+        throw new CelEvaluationError(`invalid_argument: ${value}`)
+      }
+      mapExpression[key] = value
+    }
+
+    if (!ctx.identifierDotExpression && !ctx.identifierIndexExpression) {
+      return mapExpression
+    }
+
+    return this.getIndexSection(ctx, mapExpression)
+  }
+
+  private getIndexSection(
+    ctx: MapExpressionCstChildren | IdentifierExpressionCstChildren,
+    mapExpression: unknown
+  ) {
+    const expressions = [
+      ...(ctx.identifierDotExpression || []),
+      ...(ctx.identifierIndexExpression || []),
+    ].sort((a, b) => (getPosition(a) > getPosition(b) ? 1 : -1))
+
+    return expressions.reduce((acc: unknown, expression) => {
+      if (expression.name === 'identifierDotExpression') {
+        return this.getIdentifier(acc, expression.children.Identifier[0].image)
+      }
+
+      const index = this.visit(expression.children.expr[0])
+      return this.getIdentifier(acc, index)
+    }, mapExpression)
+  }
+
+  mapKeyValues(children: MapKeyValuesCstChildren): [string, unknown] {
+    const key = this.visit(children.key)
+    const value = this.visit(children.value)
+    return [key, value]
+  }
+
   macrosExpression(ctx: MacrosExpressionCstChildren): unknown {
     const macrosIdentifier = ctx.MacrosIdentifier[0]
     // eslint-disable-next-line sonarjs/no-small-switch
     switch (macrosIdentifier.image) {
-      case 'size':
-        return ctx.arg ? this.visit(ctx.arg).length : 0
+      case 'size': // todo type it
+        return ctx.arg ? size(this.visit(ctx.arg)) : 0
```

Comment:
NIT: our `size()` returns `0`, the original one throws an error. Also not a blocker